src/helpers/vyos-domain-resolver.py

```diff
@@ -0,0 +1,131 @@
+from vyos.firewall import fqdn_update
+from vyos.util import call
+from vyos.util import commit_in_progress
+from vyos.util import dict_search_args
+from vyos.xml import defaults
+
+base = ['firewall']
+timeout = 300
+cache = False
+
+domain_state = {}
+
+def get_config(conf):
+    firewall = conf.get_config_dict(base, key_mangling=('-', '_'), get_first_key=True,
+                                    no_tag_node_value_mangle=True)
+
+    default_values = defaults(base)
+    for tmp in ['name', 'ipv6_name']:
+        if tmp in default_values:
+            del default_values[tmp]
+
+    if 'zone' in default_values:
+        del default_values['zone']
+
+    firewall = dict_merge(default_values, firewall)
+
+    global timeout, cache
+
+    if 'resolver_interval' in firewall:
+        timeout = int(firewall['resolver_interval'])
+
+    if 'resolver_cache' in firewall:
+        cache = True
+
+    fqdn_config_parse(firewall)
+
+    return firewall
+
+def resolve(domains, ipv6=False):
+    global domain_state
+
+    ip_list = set()
+
+    for domain in domains:
+        resolved = fqdn_resolve(domain, ipv6=ipv6)
+
+        if resolved and cache:
+            domain_state[domain] = resolved
+        elif not resolved:
+            if domain not in domain_state:
+                continue
+            resolved = domain_state[domain]
+
+        ip_list = ip_list | resolved
+    return ip_list
+
+def update(firewall):
+    count = 0
+    domain_groups = dict_search_args(firewall, 'group', 'domain_group')
+    if domain_groups:
+        for set_name, domain_config in domain_groups.items():
+            if 'address' not in domain_config:
+                continue
+
+            domains = domain_config['address']
+            
+            ip_list = resolve(domains, ipv6=False)
+            fqdn_update('ip mangle', f'D_{set_name}', ip_list)
+            fqdn_update('ip vyos_filter', f'D_{set_name}', ip_list)
+            fqdn_update('ip vyos_nat', f'D_{set_name}', ip_list)
+
+            ip6_list = resolve(domains, ipv6=True)
+            fqdn_update('ip6 mangle', f'D_{set_name}', ip6_list)
+            fqdn_update('ip6 vyos_filter', f'D_{set_name}', ip6_list)
+            count += 1
+
+    for set_name, domain in firewall['ip_fqdn'].items():
+        ip_list = resolve([domain], ipv6=False)
+        fqdn_update('ip vyos_filter', f'FQDN_{set_name}', ip_list)
+        count += 1
+
+    for set_name, domain in firewall['ip6_fqdn'].items():
+        ip_list = resolve([domain], ipv6=True)
+        fqdn_update('ip6 vyos_filter', f'FQDN_{set_name}', ip_list)
+        count += 1
+
+    print(f'[{time.ctime()}] Updated {count} sets')
+
+if __name__ == '__main__':
+    print(f'[{time.ctime()}] VyOS domain resolver')
+
+    count = 1
+    while commit_in_progress():
+        if ( count <= 40 ):
+            print(f'Waiting for commit in progress... {count}')
+        else:
+            print(f'Commit still in progress after {count}, continuing anyway')
```

Comment:
What consequences can be in this case? Maybe firewall structures will not be ready yet, or something similar?

Reply:
Should not we disallow this instead of skipping, in such a case?